### PR TITLE
make bitset use popcount

### LINKED ISSUE #667: <bitset>: count() should use the same approach as std::popcount
`bitset<N>::count()` 

https://github.com/microsoft/STL/blob/8e8453c67723cc28a2773fbb8eacd5821f059386/stl/inc/bitset#L341-L366

should use the same compiler magic:  as `<bit>` `popcount(x)` when it is available:

https://github.com/microsoft/STL/blob/8e8453c67723cc28a2773fbb8eacd5821f059386/stl/inc/bit#L143-L150

Blocked by the same as #313 .

Disabled here:

https://github.com/microsoft/STL/blob/8e8453c67723cc28a2773fbb8eacd5821f059386/stl/inc/yvals_core.h#L1091-L1096

Also tracked by DevCom-967643 and Microsoft-internal VSO-1090007 / [AB#1090007](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1090007).

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Resolve #667

### CODE REVIEW

**Comment by AlexGuteniev** [comment]
tr1/bitset test has enough coverage

File: stl/inc/bitset [hunk 1]
```diff
@@ -336,29 +337,36 @@ public:
...
+                } else {
+                    static_assert(_Always_false<_Ty>, "unexpected underlying type size");
+                }
+            }
+            return _Val;
+        }
+#endif // defined(_M_IX86) || defined(_M_X64)
+        size_t _Val = 0;
+        for (size_t _Wpos = 0; _Wpos <= _Words; ++_Wpos) {
+            _Val += _Popcount_fallback(_Array[_Wpos]);
```
Comment (AlexGuteniev) [inline stl/inc/bitset:368]: It may be worse than before change, especially on 32-bit CPU with 64-bit underlying type

File: stl/inc/bitset [hunk 2]
```diff
@@ -336,29 +337,36 @@ public:
...
-                                         "\2\3\3\4\3\4\4\5\3\4\4\5\4\5\5\6"
-                                         "\3\4\4\5\4\5\5\6\4\5\5\6\5\6\6\7"
-                                         "\3\4\4\5\4\5\5\6\4\5\5\6\5\6\6\7"
-                                         "\4\5\5\6\5\6\6\7\5\6\6\7\6\7\7\x8";
-        const unsigned char* _Ptr       = &reinterpret_cast<const unsigned char&>(_Array);
-        const unsigned char* const _End = _Ptr + sizeof(_Array);
-        size_t _Val                     = 0;
-        for (; _Ptr != _End; ++_Ptr) {
-            _Val += _Bitsperbyte[*_Ptr];
+#if defined(_M_IX86) || defined(_M_X64)
```
Comment (StephanTLavavej) [inline stl/inc/bitset:340]: Why does this contain the extensive logic needed to directly call the intrinsics, instead of calling `_Popcount()`?

https://github.com/microsoft/STL/blob/f75c7f596c7b491168fefb5eff5164ab7ae36867/stl/inc/limits#L1136-L1147

If the rationale is to avoid repeated tests of `_Definitely_have_popcnt`, I'd recommend profiling the difference. I expect that the test has near-zero cost, as it doesn't invoke any separately compiled functions, and will be perfectly predictable after the first time.

If it has an observable cost, I think it could be better to explore enhancing the `_Popcount()` machinery so we can lift out the check without duplicating the rest of the logic.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/bitset (hunk 2 above)
Comment (AlexGuteniev) [reply stl/inc/bitset:340]:
```C++
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

#include <bit>
#include <chrono>
#include <iostream>
#include <random>

const auto length = 10'0000;
const auto repeat = 100;

volatile int isa_available = 6;

template<std::size_t Words>
struct bitset {

    void populate() {
        std::mt19937 mt(43223);
        std::uniform_int_distribution<Ty> dis(std::numeric_limits<Ty>::min(), std::numeric_limits<Ty>::max());
        for (std::size_t Wpos = 0; Wpos <= Words; ++Wpos) {
            Array[Wpos] = dis(mt);
        }
    }

    using Ty = std::size_t;

    size_t count1() const noexcept { // count number of set bits
        Ty Val = 0;
        for (std::size_t Wpos = 0; Wpos <= Words; ++Wpos) {
            Val += std::_Popcount(Val);
        }
        return Val;
    }

    size_t count2() const noexcept { // count number of set bits
        const bool Definitely_have_popcnt = isa_available >= 5;
        if (Definitely_have_popcnt) {
            Ty Val = 0;
            for (std::size_t Wpos = 0; Wpos <= Words; ++Wpos) {
                if constexpr (sizeof(Ty) == 4) {
                    Val += __popcnt(Array[Wpos]);
                }
                else if constexpr (sizeof(Ty) == 8) {
#ifdef _M_IX86
                    _Val += (__popcnt(_Array[_Wpos] >> 32) + __popcnt(static_cast<unsigned int>(_Array[_Wpos])));
#else // ^^^ _M_IX86 / !_M_IX86 vvv
                    Val += __popcnt64(Array[Wpos]);
#endif // _M_IX86
                }
            }
            return Val;
        }

        Ty Val = 0;
        for (std::size_t Wpos = 0; Wpos <= Words; ++Wpos) {
            Val += std::_Popcount_fallback(Array[Wpos]);

        }
        return Val;
    }
    Ty Array[Words+1];
};

volatile std::size_t result;

int main()
{
    const int N = 200000;
    bitset<10'000> bs;
    bs.populate();

    std::chrono::steady_clock::duration d1{};
    std::chrono::steady_clock::duration d2{};
    std::chrono::steady_clock::time_point t;

    t = std::chrono::steady_clock::now();
    for (int i = 0; i < N; ++i) {
        result = bs.count1();
    }
    d1 = std::chrono::steady_clock::now() - t;

    t = std::chrono::steady_clock::now();
    for (int i = 0; i < N; ++i) {
        result = bs.count2();
    }
    d2 = std::chrono::steady_clock::now() - t;

    std::cout << std::chrono::duration_cast<std::chrono::duration<double>>(d1) << " vs " <<
        std::chrono::duration_cast<std::chrono::duration<double>>(d2) << "\n";
    return 0;
}
```

Comment (AlexGuteniev) [reply stl/inc/bitset:340]: 2.54269s vs 1.52024s

Comment (AlexGuteniev) [reply stl/inc/bitset:340]: So it makes sense to move dispatch out of loop.
I prefer #2079 and #2127 to land then before doing changes in `std::_Popcount`
Moving to draft for now

Comment (AlexGuteniev) [reply stl/inc/bitset]: Found a way to move the check out of loop with minimal `_Popcount` changes. See it working https://godbolt.org/z/EejPGYhr7

File: stl/inc/bitset (hunk 1 above)
Comment (AlexGuteniev) [reply stl/inc/bitset]: If it is, let's fix centralized later

File: stl/inc/bitset [hunk 3]
```diff
@@ -336,29 +337,12 @@ public:
...
-                                         "\2\3\3\4\3\4\4\5\3\4\4\5\4\5\5\6"
-                                         "\3\4\4\5\4\5\5\6\4\5\5\6\5\6\6\7"
-                                         "\3\4\4\5\4\5\5\6\4\5\5\6\5\6\6\7"
-                                         "\4\5\5\6\5\6\6\7\5\6\6\7\6\7\7\x8";
-        const unsigned char* _Ptr       = &reinterpret_cast<const unsigned char&>(_Array);
-        const unsigned char* const _End = _Ptr + sizeof(_Array);
-        size_t _Val                     = 0;
-        for (; _Ptr != _End; ++_Ptr) {
-            _Val += _Bitsperbyte[*_Ptr];
+        // local copy allows the compiler to move the condition out of loop
```
Comment (miscco) [inline stl/inc/bitset:340]: This could be `constexpr`

File: stl/inc/limits [hunk 4]
```diff
@@ -1093,10 +1093,10 @@ _NODISCARD int _Checked_x86_x64_countr_zero(const _Ty _Val) noexcept {
 
 #if _HAS_POPCNT_INTRINSICS
 template <class _Ty>
-_NODISCARD int _Checked_x86_x64_popcount(const _Ty _Val) noexcept {
+_NODISCARD int _Checked_x86_x64_popcount(const _Ty _Val, const int _Isa_available) noexcept {
```
Comment (miscco) [inline stl/inc/limits:1096]: We could pass this as a template parameter and use `if constexpr`

File: stl/inc/limits [hunk 5]
```diff
@@ -1133,14 +1133,23 @@ _NODISCARD constexpr int _Countr_zero(const _Ty _Val) noexcept {
     return _Countr_zero_fallback(_Val);
 }
 
+_NODISCARD _CONSTEXPR20 inline int _Get_isa_available() noexcept {
```
Comment (miscco) [inline stl/inc/limits:1136]: This could be unconditionally `constexpr`

File: stl/inc/limits (hunk 4 above)
Comment (AlexGuteniev) [reply stl/inc/limits:1096]: No. `__isa_available` is runtime detected.

File: stl/inc/bitset (hunk 3 above)
Comment (AlexGuteniev) [reply stl/inc/bitset:340]: No. __isa_available is runtime detected.

Comment (AlexGuteniev) [reply stl/inc/bitset:340]: and the whole `count()` is not `constexpr` in standard

File: stl/inc/limits (hunk 5 above)
Comment (AlexGuteniev) [reply stl/inc/limits]: No, it can't. In pre-C++20, when `is_constant_evaluated()` is not available, we don't have compile-time fallback, and always return `__isa_available`, so it is never constexpr.

File: stl/inc/bitset [hunk 6]
```diff
@@ -406,12 +390,12 @@ public:
 
         constexpr bool _No_padding = _Bits % _Bitsperword == 0;
         for (size_t _Wpos = 0; _Wpos < _Words + _No_padding; ++_Wpos) {
-            if (_Array[_Wpos] != ~static_cast<_Ty>(0)) {
+            if (_Array[_Wpos] != ~_Ty{0}) {
```
Comment (AlexGuteniev) [inline stl/inc/bitset:393]: Unrelated change for consistency. In other similar places there is `{ }`. The inconsistency was introduced by me a year ago.

**Comment by AlexGuteniev** [comment]
The benchmark that compares old implementation, new implementation and fallback of new implementation:
```C++
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

#include <bit>
#include <chrono>
#include <iostream>
#include <random>

int get_isa_available() noexcept {
    if (_STD is_constant_evaluated()) {
        return 0;
    }
#if defined(_M_IX86) || defined(_M_X64)
    return std::__isa_available;
#else
    return 0;
#endif
}

template <class _Ty>
int checked_x86_x64_popcount(const _Ty _Val, const int _Isa_available) noexcept {
    constexpr int _Digits = std::numeric_limits<_Ty>::digits;
#ifndef __AVX__
    const bool _Definitely_have_popcnt = _Isa_available >= __ISA_AVAILABLE_SSE42;
    if (!_Definitely_have_popcnt) {
        return std::_Popcount_fallback(_Val);
    }
#else // ^^^ pre-__AVX__ ^^^ / vvv __AVX__ vvv
    (void)_Isa_available;
#endif // ^^^ __AVX__ ^^^
    if constexpr (_Digits <= 16) {
        return static_cast<int>(__popcnt16(_Val));
    }
    else if constexpr (_Digits == 32) {

        return static_cast<int>(__popcnt(_Val));

    }
    else {

#ifdef _M_IX86

        return static_cast<int>(__popcnt(_Val >> 32) + __popcnt(static_cast<unsigned int>(_Val)));

#else // ^^^ _M_IX86 / !_M_IX86 vvv

        return static_cast<int>(__popcnt64(_Val));

#endif // _M_IX86

    }

}

template <class _Ty, std::enable_if_t<std::_Is_standard_unsigned_integer<_Ty>, int> _Enabled = 0>
int popcount(const _Ty _Val, const int _Isa_available = _Get_isa_available()) noexcept {
    if (!_STD is_constant_evaluated())
    {
        return checked_x86_x64_popcount(_Val, _Isa_available);
    }
    return std::_Popcount_fallback(_Val);
}

template<std::size_t Words>
struct bitset {

    void populate() {
        std::mt19937 mt(43223);
        std::uniform_int_distribution<Ty> dis(std::numeric_limits<Ty>::min(), std::numeric_limits<Ty>::max());
        for (std::size_t Wpos = 0; Wpos <= Words; ++Wpos) {
            Array[Wpos] = dis(mt);
        }
    }

    using Ty = std::uint64_t;

    size_t count_old() const noexcept { 
        const char* const _Bitsperbyte = 
            "\0\1\1\2\1\2\2\3\1\2\2\3\2\3\3\4"
            "\1\2\2\3\2\3\3\4\2\3\3\4\3\4\4\5"
            "\1\2\2\3\2\3\3\4\2\3\3\4\3\4\4\5"
            "\2\3\3\4\3\4\4\5\3\4\4\5\4\5\5\6"
            "\1\2\2\3\2\3\3\4\2\3\3\4\3\4\4\5"
            "\2\3\3\4\3\4\4\5\3\4\4\5\4\5\5\6"
            "\2\3\3\4\3\4\4\5\3\4\4\5\4\5\5\6"
            "\3\4\4\5\4\5\5\6\4\5\5\6\5\6\6\7"
            "\1\2\2\3\2\3\3\4\2\3\3\4\3\4\4\5"
            "\2\3\3\4\3\4\4\5\3\4\4\5\4\5\5\6"
            "\2\3\3\4\3\4\4\5\3\4\4\5\4\5\5\6"
            "\3\4\4\5\4\5\5\6\4\5\5\6\5\6\6\7"
            "\2\3\3\4\3\4\4\5\3\4\4\5\4\5\5\6"
            "\3\4\4\5\4\5\5\6\4\5\5\6\5\6\6\7"
            "\3\4\4\5\4\5\5\6\4\5\5\6\5\6\6\7"
            "\4\5\5\6\5\6\6\7\5\6\6\7\6\7\7\x8";
        const unsigned char* _Ptr = &reinterpret_cast<const unsigned char&>(Array);
        const unsigned char* const _End = _Ptr + sizeof(Array);
        size_t _Val = 0;
        for (; _Ptr != _End; ++_Ptr) {
            _Val += _Bitsperbyte[*_Ptr];
        }

        return _Val;
    }

    size_t count_new() const noexcept {
        const int _Isa_available = get_isa_available();
        size_t _Val = 0;
        for (size_t _Wpos = 0; _Wpos <= Words; ++_Wpos) {
            _Val += popcount(Array[_Wpos], _Isa_available);
        }
        return _Val;
    }

    Ty Array[Words + 1];
};

volatile std::size_t result;

int main()
{
    const int N = 200000;
    bitset<10'000> bs;
    bs.populate();

    std::chrono::steady_clock::duration d1{};
    std::chrono::steady_clock::duration d2{};
    std::chrono::steady_clock::duration d3{};
    std::chrono::steady_clock::time_point t;

    t = std::chrono::steady_clock::now();
    for (int i = 0; i < N; ++i) {
        result = bs.count_old();
    }
    d1 = std::chrono::steady_clock::now() - t;

    t = std::chrono::steady_clock::now();
    for (int i = 0; i < N; ++i) {
        result = bs.count_new();
    }
    d2 = std::chrono::steady_clock::now() - t;

    std::__isa_available = 0;
    t = std::chrono::steady_clock::now();
    for (int i = 0; i < N; ++i) {
        result = bs.count_new();
    }
    d3 = std::chrono::steady_clock::now() - t;

    std::cout << "old\t" << d1 << "\n";
    std::cout << "new\t" << d2 << "\n";
    std::cout << "new fb\t" << d3 << "\n";
    return 0;
}
```

**Comment by AlexGuteniev** [comment]
x64 benchmark result:
```
old     5340497100ns
new     1556451300ns
new fb  5810661200ns
```

x86 benchmark result:
```
old     5329002100ns
new     2556231300ns
new fb  16522223900ns
```

So it works, but the fallback become worse.
Probably should consider `_Popcnt` fallback is a separate issue.
Apparently #2079 did not help much.

**Comment by AlexGuteniev** [comment]
Turns out that for the best performance, more significant changes are needed.
In particular, caching `__isa_available` to local variable makes the compiler to duplicate a loop only for small bitsets, but the effect is very significant on all sizes. 
Also looks like the current fallback in `popcount` is worse than the current `bitset::count`

I'll wait for #2127 to be merged, then will continue, as the changes going to be extensive.

**Comment by AlexGuteniev** [comment]
Re-doe in a different PR, used another approach -- please review #2201
